### PR TITLE
Copycat.Client: address transient failures with backoff-retry logic

### PR DESCRIPTION
**Pull request by vyzo**
#75

Adds backoff retry logic to client operations.

### CODE REVIEW

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala [hunk 1]
```diff
@@ -44,31 +48,150 @@ object Copycat {
...
+    private var shutdown = false
+    private var server: Option[String] = None
     private var listeners: Set[JournalListener] = Set()
+    private var stateListeners: Set[ClientStateListener] = Set()
+    private val logger = LoggerFactory.getLogger(classOf[Client])
+    private val random = new Random
+    private val timer = new Timer(true) // runAsDaemon
+ 
+    client.onStateChange(new Consumer[CopycatClient.State] {
+                              def accept(state: CopycatClient.State) {
```
Comment (bigs) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala]:
``` scala
def accept = onStateChange
```

Comment (bigs) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala]: err edited

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala [hunk 2]
```diff
@@ -44,31 +48,116 @@ object Copycat {
...
     
-    def copycat = client
+    // submit a state machine operation with retry logic to account
+    // for potentially transient client connectivity errors
+    val maxRetries = 5
+    private def submit[T](op: Operation[T], retry: Int = 0): Future[T] = {
+      val fut = FutureConverters.toScala(client.submit(op))
+      if (retry < maxRetries) {
+        fut.recoverWith { 
+          case e: Throwable =>
```
Comment (bigs) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala]: fine for now, tho we should def dive into the errors and allow some to halt progress if necessary

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala [hunk 3]
```diff
@@ -44,31 +48,116 @@ object Copycat {
...
+      logger.info("Backing off for " + delay + " ms")
+      timer.schedule(new TimerTask {
+        def run {
+          promise.complete(Try(retry))
+        }
+      }, delay)
+      promise.future
+    }
+    
+    private def onStateChange(state: CopycatClient.State) {
```
Comment (bigs) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala]: :+1:

**Comment by bigs** [comment]
lgtm :+1:

**Comment by vyzo** [comment]
I am going to rework reconnect logic in and make the retry logic aware of the current state so that it waits for recovery before retrying.

**Comment by vyzo** [comment]
Alright, some important changes:
- the client now keeps track of its state
- retries are suspended while the connection is suspended and failed early if disconnected
- add back reconnection logic, but this time handle races more gracefully

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala [hunk 4]
```diff
@@ -44,32 +48,204 @@ object Copycat {
...
+  
+  object ClientState {
+    case object Connected extends ClientState
+    case object Suspended extends ClientState
+    case object Disconnected extends ClientState
+  }
+
+  trait ClientStateListener {
+    def onStateChange(state: ClientState): Unit
+  }
```
Comment (bigs) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala:65]: this is much cleaner. excellent

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala [hunk 5]
```diff
@@ -44,32 +48,204 @@ object Copycat {
...
+
+        case (Suspended, _) =>
+          logger.info("Copycat client is suspended; delaying "+ op)
+          backoff(1).flatMap { _ =>
+            logger.info("Retrying operation " + op)
+            submit(op, retry)
+          }
+          
+        case (Connected, _) =>
+          val fut = FutureConverters.toScala(client.submit(op))
```
Comment (bigs) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala:104]: is this whole class interacted w in a single threaded context? if not, synchronized vars may be in order

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala [hunk 6]
```diff
@@ -44,32 +48,204 @@ object Copycat {
...
+          promise.complete(Try(retry))
+        }
+      }, delay)
+      promise.future
+    }
+    
+    private def onStateChange(cstate: CopycatClient.State) {
+      cstate match {
+        case CopycatClient.State.CONNECTED => 
+          state = Connected
```
Comment (bigs) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala:136]: is there a daemon thread that will hang or does this successfully terminate

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala [hunk 7]
```diff
@@ -44,32 +48,204 @@ object Copycat {
...
+          logger.info("Copycat session connected")
+          stateListeners.foreach(_.onStateChange(Connected))
+          
+        case CopycatClient.State.SUSPENDED =>
+          state = Suspended
```
Comment (bigs) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala:141]: perfect 👍🏻

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala [hunk 8]
```diff
@@ -44,32 +48,204 @@ object Copycat {
...
     
-    def copycat = client
+    private def disconnect(what: String) {
+      state = Disconnected
+      logger.info(what)
+      stateListeners.foreach(_.onStateChange(Disconnected))
+    }
+    
+    private def reconnect() {
+      def loop(address: String, retry: Int) {
```
Comment (bigs) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala:170]: maybe throw this in a start method?

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala (hunk 5 above)
Comment (vyzo) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala:104]: it is interacted in an evented context with submits originating in outside threads and entering the event loop through the copycat completable futures.
I don't think we need to synchronize any variables, the races with shuting down are resolved with the backoff retries.

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala (hunk 8 above)
Comment (bigs) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala:170]: maybe throw this in a start method?

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala [hunk 9]
```diff
@@ -44,32 +48,204 @@ object Copycat {
...
+  
+  object ClientState {
+    case object Connected extends ClientState
+    case object Suspended extends ClientState
+    case object Disconnected extends ClientState
+  }
+
+  trait ClientStateListener {
+    def onStateChange(state: ClientState): Unit
+  }
```
Comment (bigs) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala:64]: 👍🏻

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala [hunk 10]
```diff
@@ -44,32 +48,204 @@ object Copycat {
...
+                              }
+    })
+    
+    // submit a state machine operation with retry logic to account
+    // for potentially transient client connectivity errors
+    val maxRetries = 5
+    private def submit[T](op: Operation[T], retry: Int = 0): Future[T] = {
+      (state, shutdown) match {
+        case (_, true) =>
+          Future {throw new ClientException("Copycat client has shutdown")}
```
Comment (bigs) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala:91]: maybe join all of these

**Comment by bigs** [comment]
only minor comments. overall LGTM

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala (hunk 6 above)
Comment (vyzo) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala:136]: the timer has a daemon thread, but it doesn't hang -- it passes control to the other threads through the future pipeline by completing the promise.

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala (hunk 8 above)
Comment (vyzo) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala:170]: We have connect which the user needs to call to initiaze the client.

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala (hunk 10 above)
Comment (vyzo) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala:91]: ah yes, I'll move the maxRetries up with the other variables.

Comment (vyzo) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala:91]: and cleaned up the whitespace to be more condensed.

File: transactor/src/main/scala/io/mediachain/transactor/Copycat.scala (hunk 6 above)
Comment (bigs) [inline transactor/src/main/scala/io/mediachain/transactor/Copycat.scala:136]: cool :+1:

**Pull request merged**